src/com/spike/TypeChecker.java

```diff
@@ -0,0 +1,528 @@
+      new ArrayList<>()
+    ));
+  }
+
+  private void gatherGlobals() {
+    for (Stmt stmt : stmts) {
+      if (stmt instanceof Stmt.Function) {
+        Stmt.Function function = (Stmt.Function) stmt;
+        if (function.name.lexeme.equals("main"))
+          checkMainDecl(function);
+        else
+          checkFuncDecl(function);
+      } else if (stmt instanceof Stmt.Var) {
+        Stmt.Var v = (Stmt.Var) stmt;
+        checkVarDecl(v);
+      } else {
+        reporter.report("Only variable and function declarations are allowed on top level " + stmt);
+      }
+    }
+
+    if (!hasMain) {
+      reporter.report("'main' method is not defined");
+    }
+  }
+
+  private void checkFuncDecl(Stmt.Function function) {
+    if (environment.probe(function.name)) {
+      error(function.name, "'" + function.name.lexeme + "' is already defined"); // #TODO Add prev declaration
+    } else {
+      checkSignature(function.name, function.params, function.types, function.returnType);
+      environment.define(function.name.lexeme, function);
+    }
+  }
+
+  private void checkVarDecl(Stmt.Var variable) {
+    checkType(variable.type);
+  }
+
+  private void checkMainDecl(Stmt.Function main) {
+    if (hasMain) {
+      Stmt.Function prevMain = (Stmt.Function) environment.get(main.name);
+      error(main.name, "'main' is already defined, previous declaration at main:"
+        + prevMain.name.line + ":" + prevMain.name.column);
+    } else {
+      if (main.params.size() != 0)
+        error(main.name, "'main' takes zero arguments, but " + main.params.size() + " found");
+      environment.define(main.name.lexeme, main);
+      hasMain = true;
+    }
+  }
+
+  private void checkSignature(Token functionName, List<Token> params, List<Token> types, Token returnType) {
+    beginScope();
+
+    checkTypes(types);
+    checkType(returnType);
+
+    for (Token param : params) {
+      if (environment.probe(param)) {
+        Token prev = (Token) environment.get(param);
+        error(param, "Parameter '" + param.lexeme + "' is already defined at "
+          + functionName.lexeme + ":" + prev.line + ":" + prev.column);
+      } else {
+        environment.define(param.lexeme, param);
+      }
+    }
+
+    endScope();
+  }
+
+  private void checkStmts() {
+    for (Stmt stmt : stmts) {
+      if (stmt instanceof Stmt.Var) { // FIXME: avoid bindings inside functions problems
+        stmt.accept(this);
+      }
+    }
+
+    for (Stmt stmt : stmts) {
+      if (!(stmt instanceof Stmt.Var)) {
+        stmt.accept(this);
+      }
+    }
+  }
+
+  @Override
+  public Void visit(Stmt.Var stmt) {
+    if (environment.probe(stmt.name)) {
+      error(stmt.name, "'" + stmt.name.lexeme + "' is already defined"); // #TODO Add prev declaration
+    } else {
+      environment.define(stmt.name.lexeme, stmt.type);
+      int type = (stmt.init == null) ? -1 : stmt.init.accept(this);
+      if (type != -1 && type != classTable.get(stmt.type.lexeme)) {
```

Comment:
Not sure but this seems like a `NullPointerException` if `ClassTable` doesn't have `stmt.type.lexeme`

Reply:
Maybe. I've checked that types exist for top-level variables and parameters in function declaration. But theoretically (should check) the following could produce `NullPointerException`
'''
def main() void {
    var x: UndefinedType; // error
}
'''